core/src/Cuda/Kokkos_Cuda_Instance.cpp

```diff
@@ -438,47 +446,49 @@ Kokkos::Cuda::initialize WARNING: Cuda is allocating into UVMSpace by default
 
 //----------------------------------------------------------------------------
 
-using ScratchGrain = Cuda::size_type[Impl::CudaTraits::WarpSize];
-enum { sizeScratchGrain = sizeof(ScratchGrain) };
-
 Cuda::size_type *CudaInternal::scratch_flags(const std::size_t size) const {
   if (verify_is_initialized("scratch_flags") &&
-      m_scratchFlagsCount * sizeScratchGrain < size) {
-    m_scratchFlagsCount = (size + sizeScratchGrain - 1) / sizeScratchGrain;
+      m_scratchFlagsCount < scratch_count(size)) {
+    m_scratchFlagsCount = scratch_count(size);
 
     using Record =
         Kokkos::Impl::SharedAllocationRecord<Kokkos::CudaSpace, void>;
 
     if (m_scratchFlags) Record::decrement(Record::get_record(m_scratchFlags));
 
-    Record *const r =
-        Record::allocate(Kokkos::CudaSpace(), "Kokkos::InternalScratchFlags",
-                         (sizeof(ScratchGrain) * m_scratchFlagsCount));
+    std::size_t alloc_size;
+    if (multiply_overflow(m_scratchFlagsCount, sizeScratchGrain, alloc_size)) {
+      Kokkos::abort("Arithmetic overflow detected.");
+    }
```

Comment:
I don't find this code intuitive. What about we do sth like
```c++
std::size_t alloc_size = multiply_overflow(m_scratchFlagsCount, sizeScratchGrain);
```
and we put the `Kokkos::abort` inside `multiply_overflow`. The code is easier to read and we don't need to repeat the `Kokkos::abort` everywhere.

Reply:
See ffe1120.